dali/kernels/common/simd.h

```diff
@@ -0,0 +1,209 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_COMMON_SIMD_H_
+#define DALI_KERNELS_COMMON_SIMD_H_
+
+#ifdef __SSE2__
+#include <emmintrin.h>
+#endif
+
+#include <cstddef>
+#include <type_traits>
+#include "dali/core/force_inline.h"
+
+namespace dali {
+namespace kernels {
+namespace simd {
+
+#ifdef __SSE2__
+
+template <int n>
+struct float4x {
+  __m128 v[n];  // NOLINT
+};
+
+template <int n>
+struct i128x {
+  __m128i v[n];  // NOLINT
+};
+
+using float4x1 = float4x<1>;
+using float4x2 = float4x<2>;
+using float4x4 = float4x<4>;
+using i128x1 = i128x<1>;
+using i128x2 = i128x<2>;
+using i128x4 = i128x<4>;
+
+inline __m128i clamp_round(__m128 f, float lo, float hi) {
+  f = _mm_min_ps(_mm_max_ps(f, _mm_set1_ps(lo)), _mm_set1_ps(hi));
+  return _mm_cvtps_epi32(f);  // round
+}
+
+inline void store(int8_t *i8, i128x4 iv) {
+  __m128i sv0 = _mm_packs_epi32(iv.v[0], iv.v[1]);
+  __m128i sv1 = _mm_packs_epi32(iv.v[2], iv.v[3]);
+  _mm_storeu_si128(reinterpret_cast<__m128i*>(i8), _mm_packs_epi16(sv0, sv1));
+}
+
+inline void store(uint8_t *u8, i128x4 iv) {
+  __m128i sv0 = _mm_packs_epi32(iv.v[0], iv.v[1]);
+  __m128i sv1 = _mm_packs_epi32(iv.v[2], iv.v[3]);
+  _mm_storeu_si128(reinterpret_cast<__m128i*>(u8), _mm_packus_epi16(sv0, sv1));
+}
+
+inline void store(int16_t *i16, i128x2 iv) {
+  _mm_storeu_si128(reinterpret_cast<__m128i*>(i16), _mm_packs_epi32(iv.v[0], iv.v[1]));
+}
+
+inline void store(uint16_t *u16, i128x2 iv) {
+  __m128i sh = _mm_set_epi32(0, 0, 0, 1);
+  __m128i one = _mm_set1_epi32(1);
+  // avoid signed saturation by shifting right and saving LSB, then reconsitituting it
+  __m128i hi0 = _mm_srl_epi32(iv.v[0], sh);
+  __m128i hi1 = _mm_srl_epi32(iv.v[1], sh);
+  __m128i lo0 = _mm_and_si128(iv.v[0], one);
+  __m128i lo1 = _mm_and_si128(iv.v[0], one);
+  __m128i lo = _mm_packs_epi32(lo0, lo1);
+  __m128i hi = _mm_packs_epi32(hi0, hi1);
+  __m128i out = _mm_or_si128(_mm_sll_epi16(hi, sh), lo);
+  _mm_storeu_si128(reinterpret_cast<__m128i*>(u16), out);
+}
+
+inline float4x4 load_f(const uint8_t *u8) {
+  __m128i in = _mm_loadu_si128(reinterpret_cast<const __m128i *>(u8));
+  __m128i zero = _mm_setzero_si128();
+  __m128i lo16 = _mm_unpacklo_epi8(in, zero);
+  __m128i hi16 = _mm_unpackhi_epi8(in, zero);
+  __m128i i32_0 = _mm_unpacklo_epi8(lo16, zero);
+  __m128i i32_1 = _mm_unpackhi_epi8(lo16, zero);
+  __m128i i32_2 = _mm_unpacklo_epi8(hi16, zero);
+  __m128i i32_3 = _mm_unpackhi_epi8(hi16, zero);
+  return {{ _mm_cvtepi32_ps(i32_0),
+            _mm_cvtepi32_ps(i32_1),
+            _mm_cvtepi32_ps(i32_2),
+            _mm_cvtepi32_ps(i32_3) }};
+}
+
+inline __m128i sext8_32(__m128i i) {
+  __m128i sh = _mm_set_epi32(0, 0, 0, 24);
+  return _mm_sra_epi32(_mm_sll_epi32(i, sh), sh);
+}
+
+inline float4x4 load_f(const int8_t *i8) {
```

Comment:
Loads a vector of int8 and expands it to vectors of float.

Reply:
You can add a comment here about it.